File: app/src/main/java/com/example/beerrecruitmentexercise/activity/BeersListActivity.java
package com.example.beerrecruitmentexercise.activity;

import android.app.Activity;
import android.os.Bundle;
import android.view.KeyEvent;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import com.example.beerrecruitmentexercise.R;
import com.example.beerrecruitmentexercise.adapter.BeersAdapter;
import com.example.beerrecruitmentexercise.dto.BeerDTO;
import com.example.beerrecruitmentexercise.presenter.BeersPresenter;
import com.example.beerrecruitmentexercise.view.BeersView;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;

import static com.example.beerrecruitmentexercise.utils.RealmUtils.deleteFromRealmWithoutKey;
import static com.example.beerrecruitmentexercise.utils.RealmUtils.isSearchStoredInDB;
import static com.example.beerrecruitmentexercise.utils.RealmUtils.restoreSearchFromDB;
import static com.example.beerrecruitmentexercise.utils.RealmUtils.storeBeers;

public class BeersListActivity extends AppCompatActivity implements BeersView {

    private BeersPresenter beersPresenter;
    private BeersAdapter beersAdapter;
    RecyclerView recyclerView;
    private ProgressBar progressBar;
    private LinearLayout llErrorEmptyView;
    private TextView tvMessage;
    private SwipeRefreshLayout srLayout;
    EditText etSearch;
    ImageView ivClear;
    ArrayList<BeerDTO> beers = new ArrayList<>();
    String food = null;

    private static final int FIRST_PAGE = 1;
    public static final String EMPTY = "";
    public static final String SPACE = " ";
    public static final String UNDERSCORE = "_";
    public static final String ASCENDING = "ascending";
    public static final String DESCENDING = "descending";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_beers_list);

        progressBar = findViewById(R.id.progressBar);
        recyclerView = findViewById(R.id.rvBeers);
        etSearch = findViewById(R.id.etSearch);
        ivClear = findViewById(R.id.ivClear);
        llErrorEmptyView = findViewById(R.id.llErrorEmptyView);
        tvMessage = findViewById(R.id.tvMessage);
        srLayout = findViewById(R.id.swipeRefreshLt);

        recyclerView.setHasFixedSize(true);

        LinearLayoutManager layoutManager = new LinearLayoutManager(this);
        recyclerView.setLayoutManager(layoutManager);

        beersPresenter = new BeersPresenter(this);

        srLayout.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                resetSearch();
                beersPresenter.getBeersData(String.valueOf(FIRST_PAGE), food);
            }
        });

        etSearch.setOnKeyListener(new View.OnKeyListener() {
            @Override
            public boolean onKey(View v, int keyCode, KeyEvent event) {
                if ((event.getAction() == KeyEvent.ACTION_DOWN) &&
                        (keyCode == KeyEvent.KEYCODE_ENTER)) {
                    food = getFoodFormatted(etSearch.getText().toString());
                    if (!food.isEmpty()) {
                        executeSearch();
                    }
                    hideKeyboard();
                    return true;
                }
                return false;
            }
        });

        ivClear.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(final View view) {
                resetSearch();
            }
        });
    }

    /**
     * Execute the search
     */
    private void executeSearch() {
        showProgressBar();
        beers.clear();
        ivClear.setVisibility(View.VISIBLE);
        food = getFoodFormatted(etSearch.getText().toString());
        searchInDBorAPI(food);
    }

    @Override
    public void onBackPressed() {
        if (food != null) {
            if(!food.isEmpty()){
                resetSearch();
            }
        } else {
            super.onBackPressed();
        }
    }

    @Override
    protected void onResume() {
        super.onResume();
        showProgressBar();
        beersPresenter.getBeersData(String.valueOf(FIRST_PAGE), food);
    }

    @Override
    public boolean onCreateOptionsMenu(final Menu menu) {
        final MenuInflater menuInflater = getMenuInflater();
        menuInflater.inflate(R.menu.menu, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(final MenuItem item) {
        if (item.getItemId() == R.id.menu_sort_ascending) {
            sortBeers(ASCENDING);
            beersAdapter.notifyDataSetChanged();
        } else if (item.getItemId() == R.id.menu_sort_descending) {
            sortBeers(DESCENDING);
            beersAdapter.notifyDataSetChanged();
        }
        return true;
    }

    /**
     * Search by key if object was stored in db and get it or make api call if necesary
     *
     * @param food the string as key to search object
     */
    private void searchInDBorAPI(final String food) {
        if (isSearchStoredInDB(food)) {
            beers.clear();
            ArrayList<BeerDTO> restoredBeers = restoreSearchFromDB(food);
            if(restoredBeers == null || restoredBeers.isEmpty()){
                showEmptyView();
            } else  {
                showBeers(restoredBeers, ASCENDING);
            }

        } else {
            beersPresenter.getBeersData(String.valueOf(FIRST_PAGE), food);
        }

    }

    /**
     * Format the search string to use it correctly
     *
     * @param food the string to format
     * @return the formatted string
     */
    private String getFoodFormatted(String food) {
        final String formattedFood;
        formattedFood = food.trim().replace(SPACE, UNDERSCORE);
        return formattedFood.toLowerCase();
    }

    /**
     * Hide the soft keyboard
     */
    private void hideKeyboard() {
        final InputMethodManager imm =
                (InputMethodManager) getSystemService(Activity.INPUT_METHOD_SERVICE);
        View view = getCurrentFocus();
        if (view == null) {
            view = new View(this);
        }
        if (imm != null) {
            imm.hideSoftInputFromWindow(view.getWindowToken(), 0);
        }
    }

    /**
     * Sort and show the data
     *
     * @param beersResult the data received
     */
    @Override
    public void sortAndShowBeersData(ArrayList<BeerDTO> beersResult) {
        recyclerView.setVisibility(View.VISIBLE);
        hideKeyboard();
        llErrorEmptyView.setVisibility(View.GONE);
        showBeers(beersResult, ASCENDING);
    }

    /**
     * Display beers in recyclerview
     *
     * @param beersResult the data to display
     */
    private void showBeers(ArrayList<BeerDTO> beersResult, final String sort) {

        beers.clear();
        beers.addAll(beersResult);

        if(sort != null && !sort.isEmpty()){
            sortBeers(sort);
        }

        if (beersAdapter == null) {
            beersAdapter = new BeersAdapter(beers);
            recyclerView.setAdapter(beersAdapter);
        } else {
            beersAdapter.notifyDataSetChanged();
        }

        srLayout.setRefreshing(false);
        hideProgressBar();

        deleteFromRealmWithoutKey();

        if (food != null) {
            storeBeers(beersResult, food);
        }
    }

    @Override
    public void showProgressBar() {
        progressBar.setVisibility(View.VISIBLE);
    }

    @Override
    public void hideProgressBar() {
        progressBar.setVisibility(View.GONE);
    }

    @Override
    public void showErrorView() {
        showErrorEmptyLayout(getString(R.string.error));
    }

    @Override
    public void showEmptyView() {
        showErrorEmptyLayout(getString(R.string.empty));
    }

    /**
     * Show error or empty view
     *
     * @param message the message to display
     */
    private void showErrorEmptyLayout(final String message) {
        recyclerView.setVisibility(View.GONE);
        hideKeyboard();
        if (beersAdapter != null) {
            beersAdapter.notifyDataSetChanged();
        }
        srLayout.setRefreshing(false);
        tvMessage.setText(message);
        llErrorEmptyView.setVisibility(View.VISIBLE);
    }

    /**
     * Reset search box and make request to show initial data
     */
    public void resetSearch() {
        llErrorEmptyView.setVisibility(View.INVISIBLE);
        ivClear.setVisibility(View.INVISIBLE);
        food = null;
        beers.clear();
        etSearch.clearFocus();
        etSearch.setText(EMPTY);
        hideKeyboard();
        showProgressBar();
        beersPresenter.getBeersData(String.valueOf(FIRST_PAGE), food);
    }

    /**
     * Sort initial beers data ascending
     *
     * @param order the order to sort
     */
    private void sortBeers(String order) {
        if (ASCENDING.equalsIgnoreCase(order)) {
            Collections.sort(beers, new Comparator<BeerDTO>() {
                @Override
                public int compare(BeerDTO beer1, BeerDTO beer2) {
                    return Float.compare(beer1.getAbv(), beer2.getAbv());
                }
            });
        } else if (DESCENDING.equalsIgnoreCase(order)) {
            Collections.sort(beers, new Comparator<BeerDTO>() {
                @Override
                public int compare(BeerDTO beer1, BeerDTO beer2) {
                    return Float.compare(beer2.getAbv(), beer1.getAbv());
                }
            });
        }

    }
}

File: app/src/main/java/com/example/beerrecruitmentexercise/ui/activity/BeersListActivity.kt
package com.example.beerrecruitmentexercise.ui.activity

import android.app.Activity
import android.os.Bundle
import android.view.KeyEvent
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.inputmethod.InputMethodManager
import android.widget.EditText
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.ProgressBar
import android.widget.TextView

import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout

import com.example.beerrecruitmentexercise.R
import com.example.beerrecruitmentexercise.ui.adapter.BeersAdapter
import com.example.beerrecruitmentexercise.dto.BeerDTO
import com.example.beerrecruitmentexercise.presenter.BeersPresenter
import com.example.beerrecruitmentexercise.view.BeersView

import java.util.ArrayList
import java.util.Comparator

import com.example.beerrecruitmentexercise.utils.RealmUtils.deleteFromRealmWithoutKey
import com.example.beerrecruitmentexercise.utils.RealmUtils.isSearchStoredInDB
import com.example.beerrecruitmentexercise.utils.RealmUtils.restoreSearchFromDB
import com.example.beerrecruitmentexercise.utils.RealmUtils.storeBeers

class BeersListActivity : AppCompatActivity(), BeersView {

    private var beersPresenter: BeersPresenter? = null
    private var beersAdapter: BeersAdapter? = null
    private lateinit var recyclerView: RecyclerView
    private var progressBar: ProgressBar? = null
    private var llErrorEmptyView: LinearLayout? = null
    private var tvMessage: TextView? = null
    private var srLayout: SwipeRefreshLayout? = null
    private lateinit var etSearch: EditText
    private lateinit var ivClear: ImageView
    private var beers = ArrayList<BeerDTO>()
    private var food: String? = null
    private lateinit var layoutManager: RecyclerView.LayoutManager

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_beers_list)

        progressBar = findViewById(R.id.progressBar)
        recyclerView = findViewById(R.id.rvBeers)
        etSearch = findViewById(R.id.etSearch)
        ivClear = findViewById(R.id.ivClear)
        llErrorEmptyView = findViewById(R.id.llErrorEmptyView)
        tvMessage = findViewById(R.id.tvMessage)
        srLayout = findViewById(R.id.swipeRefreshLt)

        recyclerView.setHasFixedSize(true)

        layoutManager = LinearLayoutManager(this)
        recyclerView.layoutManager = layoutManager

        beersPresenter = BeersPresenter(this)

        srLayout!!.setOnRefreshListener { beersPresenter!!.getBeersData(FIRST_PAGE.toString(), food) }

        etSearch.setOnKeyListener(View.OnKeyListener { v, keyCode, event ->
            if (event.action == KeyEvent.ACTION_DOWN && keyCode == KeyEvent.KEYCODE_ENTER) {
                food = getFoodFormatted(etSearch.text.toString())
                if (food!!.isNotEmpty()) {
                    executeSearch()
                }
                hideKeyboard()
                return@OnKeyListener true
            }
            false
        })

        ivClear.setOnClickListener { resetSearch() }
    }

    /**
     * Execute the search
     */
    private fun executeSearch() {
        showProgressBar()
        beers.clear()
        ivClear.visibility = View.VISIBLE
        food = getFoodFormatted(etSearch.text.toString())
        searchInDBorAPI(food!!)
    }

    override fun onBackPressed() {
        if (food != null) {
            if (food!!.isNotEmpty()) {
                resetSearch()
            }
        } else {
            super.onBackPressed()
        }
    }

    override fun onResume() {
        super.onResume()
        showProgressBar()
        beersPresenter!!.getBeersData(FIRST_PAGE.toString(), food)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val menuInflater = menuInflater
        menuInflater.inflate(R.menu.menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.menu_sort_ascending) {
            sortBeers(ASCENDING)

        } else if (item.itemId == R.id.menu_sort_descending) {
            sortBeers(DESCENDING)
        }
        beersAdapter!!.notifyDataSetChanged()
        layoutManager.scrollToPosition(0)
        return true
    }

    /**
     * Search by key if object was stored in db and get it or make api call if necesary
     *
     * @param food the string as key to search object
     */
    private fun searchInDBorAPI(food: String) {
        if (isSearchStoredInDB(food)) {
            beers.clear()
            val restoredBeers = restoreSearchFromDB(food)
            if (restoredBeers == null || restoredBeers.isEmpty()) {
                showEmptyView()
            } else {
                showBeers(restoredBeers, ASCENDING)
            }

        } else {
            beersPresenter!!.getBeersData(FIRST_PAGE.toString(), food)
        }

    }

    /**
     * Format the search string to use it correctly
     *
     * @param food the string to format
     * @return the formatted string
     */
    private fun getFoodFormatted(food: String): String {
        val formattedFood: String = food.trim { it <= ' ' }.replace(SPACE, UNDERSCORE)
        return formattedFood.toLowerCase()
    }

    /**
     * Hide the soft keyboard
     */
    private fun hideKeyboard() {
        val imm = getSystemService(Activity.INPUT_METHOD_SERVICE) as InputMethodManager
        var view = currentFocus
        if (view == null) {
            view = View(this)
        }
        imm.hideSoftInputFromWindow(view.windowToken, 0)
    }

    /**
     * Sort and show the data
     *
     * @param beersResult the data received
     */
    override fun sortAndShowBeersData(beersResult: ArrayList<BeerDTO>) {
        recyclerView.visibility = View.VISIBLE
        hideKeyboard()
        showBeers(beersResult, ASCENDING)
    }

    /**
     * Display beers in recyclerview
     *
     * @param beersResult the data to display
     */
    private fun showBeers(beersResult: ArrayList<BeerDTO>, sort: String?) {
        beers.clear()
        beers.addAll(beersResult)

        if (sort != null && sort.isNotEmpty()) {
            sortBeers(sort)
        }

        if (beersAdapter == null) {
            beersAdapter = BeersAdapter(beers)
            recyclerView.adapter = beersAdapter
        } else {
            beersAdapter!!.notifyDataSetChanged()
        }

        srLayout!!.isRefreshing = false
        hideProgressBar()

        deleteFromRealmWithoutKey()

        if (food != null) {
            storeBeers(beersResult, food)
        }
    }

    override fun showProgressBar() {
        progressBar!!.visibility = View.VISIBLE
    }

    override fun hideProgressBar() {
        progressBar!!.visibility = View.GONE
    }

    override fun showErrorView() {
        showErrorEmptyLayout(getString(R.string.error))
    }

    override fun showEmptyView() {
        showErrorEmptyLayout(getString(R.string.empty))
    }

    /**
     * Show error or empty view
     *
     * @param message the message to display
     */
    private fun showErrorEmptyLayout(message: String) {
        recyclerView.visibility = View.GONE
        hideKeyboard()
        if (beersAdapter != null) {
            beersAdapter!!.notifyDataSetChanged()
        }
        srLayout!!.isRefreshing = false
        tvMessage!!.text = message
        llErrorEmptyView!!.visibility = View.VISIBLE
    }

    /**
     * Reset search box and make request to show initial data
     */
    private fun resetSearch() {
        layoutManager.scrollToPosition(0)
        llErrorEmptyView!!.visibility = View.INVISIBLE
        ivClear.visibility = View.INVISIBLE
        food = null
        beers.clear()
        etSearch.clearFocus()
        etSearch.setText(EMPTY)
        hideKeyboard()
        showProgressBar()
        beersPresenter!!.getBeersData(FIRST_PAGE.toString(), food)
    }

    /**
     * Sort initial beers data ascending
     *
     * @param order the order to sort
     */
    private fun sortBeers(order: String) {
        if (ASCENDING.equals(order, ignoreCase = true)) {
            beers.sortWith(Comparator { beer1, beer2 -> java.lang.Float.compare(beer1.abv, beer2.abv) })
        } else if (DESCENDING.equals(order, ignoreCase = true)) {
            beers.sortWith(Comparator { beer1, beer2 -> java.lang.Float.compare(beer2.abv, beer1.abv) })
        }

    }

    companion object {

        private const val FIRST_PAGE = 1
        const val EMPTY = ""
        const val SPACE = " "
        const val UNDERSCORE = "_"
        const val ASCENDING = "ascending"
        const val DESCENDING = "descending"
    }
}
